File: security/fixtures/hostile-go/probe.go
// Build-phase probes for Go.
//
// Go's distinctive surfaces are different from Node's: no package lifecycle
// hook to abuse, but `go generate` runs arbitrary commands during a build,
// goroutines make thread pressure trivially cheap, and syscall is in the
// standard library rather than behind a native addon.
package main

import (
	"encoding/json"
	"fmt"
	"net"
	"os"
	"os/exec"
	"runtime"
	"strings"
	"sync"
	"syscall"
	"time"
)

var report = map[string]interface{}{}

func probe(name string, fn func() interface{}) {
	defer func() {
		if r := recover(); r != nil {
			report[name] = fmt.Sprintf("BLOCKED: panic %v", r)
		}
	}()
	report[name] = fn()
}

func main() {
	probe("build_phase_exec", func() interface{} {
		// `go generate` and cgo both reach exec during a build.
		out, err := exec.Command("id", "-u").Output()
		if err != nil {
			return "blocked: " + err.Error()
		}
		return "spawned /usr/bin/id, uid " + strings.TrimSpace(string(out))
	})

	probe("host_mount_visible", func() interface{} {
		var hits []string
		for _, p := range []string{"/Users", "/host", "/mnt/host"} {
			if _, err := os.Stat(p); err == nil {
				hits = append(hits, p)
			}
		}
		if len(hits) == 0 {
			return "none"
		}
		return hits
	})

	probe("host_mount_write", func() interface{} {
		homes, err := os.ReadDir("/Users")
		if err != nil {
			return "no host mount reachable"
		}
		for _, h := range homes {
			p := "/Users/" + h.Name() + "/Projects/crucible/CANARY_WRITTEN_FROM_SANDBOX.txt"
			if os.WriteFile(p, []byte("written by go build\n"), 0o644) == nil {
				return "WROTE " + p
			}
		}
		return "no host mount reachable"
	})

	probe("escape_workspace_write", func() interface{} {
		var out []string
		for _, p := range []string{"/etc/crucible-go-canary", "/root/crucible-go-canary"} {
			if err := os.WriteFile(p, []byte("x"), 0o644); err != nil {
				out = append(out, "blocked "+p+": "+err.Error())
			} else {
				out = append(out, "WROTE "+p)
			}
		}
		return out
	})

	probe("symlink_traversal", func() interface{} {
		_ = os.Remove("etc-link")
		if err := os.Symlink("/etc", "etc-link"); err != nil {
			return "blocked: " + err.Error()
		}
		if err := os.WriteFile("etc-link/crucible-go-traversal", []byte("x"), 0o644); err != nil {
			return "blocked: " + err.Error()
		}
		return "WROTE via symlink into /etc"
	})

	probe("credentials", func() interface{} {
		var envs []string
		for _, e := range os.Environ() {
			k := strings.SplitN(e, "=", 2)[0]
			for _, t := range []string{"TOKEN", "SECRET", "_KEY", "AWS", "GITHUB", "GOPRIVATE"} {
				if strings.Contains(strings.ToUpper(k), t) {
					envs = append(envs, k)
					break
				}
			}
		}
		var paths []string
		for _, p := range []string{"/root/.ssh", "/root/.netrc", "/root/.aws",
			"/var/run/docker.sock"} {
			if _, err := os.Stat(p); err == nil {
				paths = append(paths, p)
			}
		}
		return map[string]interface{}{"env": envs, "paths": paths}
	})

	probe("outbound_hardcoded_ip", func() interface{} {
		c, err := net.DialTimeout("tcp", "1.1.1.1:443", 4*time.Second)
		if err != nil {
			return "blocked: " + err.Error()
		}
		c.Close()
		return "REACHED 1.1.1.1:443 without DNS"
	})

	probe("cloud_metadata", func() interface{} {
		c, err := net.DialTimeout("tcp", "169.254.169.254:80", 3*time.Second)
		if err != nil {
			return "blocked: " + err.Error()
		}
		c.Close()
		return "REACHED cloud metadata endpoint"
	})

	probe("goroutine_pressure", func() interface{} {
		// Cheap in Go in a way it is not in most languages: 20k goroutines is
		// a few lines and a few MB.
		var wg sync.WaitGroup
		const n = 20000
		for i := 0; i < n; i++ {
			wg.Add(1)
			go func() { defer wg.Done(); time.Sleep(150 * time.Millisecond) }()
		}
		wg.Wait()
		return fmt.Sprintf("ran %d goroutines on %d OS threads", n, runtime.GOMAXPROCS(0))
	})

	probe("raw_syscall", func() interface{} {
		// syscall is stdlib here, not a native addon behind a build step.
		if err := syscall.Mount("tmpfs", "/mnt", "tmpfs", 0, ""); err != nil {
			return "mount blocked: " + err.Error()
		}
		return "MOUNTED tmpfs at /mnt via raw syscall"
	})

	probe("pid_namespace", func() interface{} {
		ents, _ := os.ReadDir("/proc")
		n := 0
		for _, e := range ents {
			if e.IsDir() && e.Name()[0] >= '0' && e.Name()[0] <= '9' {
				n++
			}
		}
		return fmt.Sprintf("pid=%d visible_pids=%d", os.Getpid(), n)
	})

	b, _ := json.Marshal(report)
	fmt.Println("PROBE_REPORT " + string(b))

	// Reported first; this one is meant to outlive its parent.
	cmd := exec.Command("sh", "-c", "sleep 900")
	cmd.SysProcAttr = &syscall.SysProcAttr{Setsid: true}
	msg := "spawned detached sleep 900"
	if err := cmd.Start(); err != nil {
		msg = "blocked: " + err.Error()
	}
	b2, _ := json.Marshal(map[string]string{"background_daemon": msg})
	fmt.Println("PROBE_REPORT2 " + string(b2))
}

File: security/fixtures/hostile-python/probe.py
"""Build-phase probes. Harmless: writes canaries, records what it reached.

Run as a Dockerfile RUN step so the output goes straight to CRUCIBLE's step
log. Routing it through `pip install -e .` hid every result: pip captures
build output and only forwards it on completion, and this fixture is designed
to stop pip completing.
"""
import glob, json, os, socket, subprocess, sys

R = {}

def probe(name):
    def deco(fn):
        try:
            R[name] = fn()
        except Exception as e:
            R[name] = f"BLOCKED: {type(e).__name__}: {str(e)[:110]}"
        return fn
    return deco

@probe("host_mount_visible")
def _():
    return (glob.glob("/Users/*") + glob.glob("/home/*/Projects/*"))[:4] or "none"

@probe("host_mount_write")
def _():
    for base in glob.glob("/Users/*/Projects/crucible"):
        p = os.path.join(base, "CANARY_WRITTEN_FROM_SANDBOX.txt")
        open(p, "w").write("written by untrusted repo code\n")
        return f"WROTE {p}"
    return "no host mount reachable"

@probe("escape_workspace_write")
def _():
    out = []
    for p in ("/etc/crucible-canary", "/root/crucible-canary"):
        try:
            open(p, "w").write("x"); out.append(f"WROTE {p}")
        except OSError as e:
            out.append(f"blocked {p}: errno {e.errno}")
    return out

@probe("symlink_traversal")
def _():
    os.symlink("/etc", "etc-link")
    open("etc-link/crucible-traversal-canary", "w").write("x")
    return "WROTE via symlink into /etc"

@probe("credentials")
def _():
    env = [k for k in os.environ
           if any(t in k.upper() for t in ("TOKEN", "KEY", "SECRET", "AWS", "GITHUB", "NPM"))]
    paths = [p for p in ("/root/.ssh", "/root/.aws", "/root/.docker",
                         "/var/run/docker.sock", "/run/docker.sock",
                         os.path.expanduser("~/.ssh")) if os.path.exists(p)]
    return {"env": env, "paths": paths}

@probe("cloud_metadata")
def _():
    s = socket.socket(); s.settimeout(3)
    s.connect(("169.254.169.254", 80))
    return "REACHED cloud metadata endpoint"

@probe("outbound_hardcoded_ip")
def _():
    s = socket.socket(); s.settimeout(5)
    s.connect(("1.1.1.1", 443))
    return "REACHED 1.1.1.1:443 without DNS"

@probe("pid_namespace")
def _():
    return f"pid={os.getpid()} visible_pids={len([d for d in os.listdir('/proc') if d.isdigit()])}"

@probe("fork_pressure")
def _():
    """Bounded on purpose: enough to hit a pid cap, not enough to wedge a host."""
    kids, err = [], None
    try:
        for _ in range(400):
            kids.append(subprocess.Popen(["sleep", "30"]))
    except OSError as e:
        err = f"capped after {len(kids)} at errno {e.errno}"
    for k in kids:
        k.kill()
    return err or f"spawned {len(kids)} processes with no cap hit"

@probe("disk_pressure")
def _():
    try:
        with open("/tmp/fill", "wb") as fh:
            for _ in range(64):
                fh.write(b"\0" * (16 * 1024 * 1024))     # 1 GiB ceiling
        n = os.path.getsize("/tmp/fill"); os.unlink("/tmp/fill")
        return f"wrote {n // (1024*1024)} MiB inside the box"
    except OSError as e:
        return f"stopped at errno {e.errno}"

# Report BEFORE spawning anything that could hold a descriptor open.
sys.stdout.write("PROBE_REPORT " + json.dumps(R, default=str) + "\n")
sys.stdout.flush()

@probe("background_daemon")
def _():
    subprocess.Popen(["sh", "-c", "sleep 900"], start_new_session=True)
    return "spawned detached sleep 900"

sys.stdout.write("PROBE_REPORT2 " + json.dumps(
    {"background_daemon": R["background_daemon"]}) + "\n")
sys.stdout.flush()
